### PR TITLE
[orc8r][directoryd] Add hwid to directory record ids state indexing.

### LINKED ISSUE #5613: Add gateway to subscriber routes and backend to Orchestrator API

### PR DESCRIPTION
**Pull request by mpgermano**
Signed-off-by: Michael Germano <mgermano@fb.com>

## Summary

This PR adds support for lookup of all subscribers on a given gateway. This is done by indexing directoryd records
to aggregate hardware ID -> record IDs. A stacked PR implements the associated API handlers to expose this 
functionality to the NMS.

Note: Given the protos don't support a map of strings to array, the decision was made to create a custom message type for DirectoryRecordIDs. The alternative here would be to map hwid's to byte blobs. This method was avoided as it forces clients to be responsible for uniform de/serialization. 

## Test Plan

Added unit tests for:
1. storage
2. storage integ test
3. directoryd servicers
4. directoryd client api
5. indexer servicer

Tested locally via a custom CLI.

### CODE REVIEW

File: orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go]: Can you add this block to a separate helper fn -- ideally indexImpl should be as minimal as possible

File: orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go]: Use multierror pkg for this, we already use it in e.g. Andy's swagger code

File: orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go]: "mergeStateErrs"

Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go]: Make input variadic, to support adding future logical indexers

File: orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go:138]: Even if we leave the code generic, can you add comment explaining why DeviceID contains the IMSI, either here or higher in the callstack

File: orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/servicers/indexer_servicer.go]: "stateErrs"

File: orc8r/protos/directoryd.proto
Comment (hcgatewood) [inline orc8r/protos/directoryd.proto:53]: Yeah I cry a bit that we can't do map<string, array[type]>, but alas

File: orc8r/cloud/go/services/directoryd/client_api.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/client_api.go]: - Nit here and elsewhere: no need to include "failed to" -- it's implicit at this level in the callstack/errorstack 
- Generally: if you're interested, I've settled on "error getting ..." when it really feels like I need to make it clear there was an issue

File: orc8r/cloud/go/services/directoryd/client_api.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/client_api.go]: Consider making explicit helper functions for these conversions between proto and native types

File: orc8r/cloud/go/services/directoryd/types/types.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/types/types.go:88]: Consider `(string, bool)` return type, is easier to semantically deal with "not found" than it is to determine black-box whether an err means "not found" vs. "bad shit"

File: orc8r/cloud/go/services/directoryd/storage/storage.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/storage/storage.go]: "by a hardware ID"

File: orc8r/cloud/go/services/directoryd/storage/storage_blobstore_test.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/storage/storage_blobstore_test.go:319]: Side note: I've fallen out of love with these impl-specific tests (pass-through tests). They take a lot of time to write, lot of time to update, and I'm unconvinced they provide much value on top of the integ tests (interface-level tests). So def fine to leave this but generally speaking I don't think it's worth the time

*One asterisk here though is for critical storages like certifier, accessd, configurator, etc, I think it makes sense to double-test. But for something like this, as long as it passes the iface-level integ test I think it's fine

File: orc8r/cloud/go/services/directoryd/storage/storage_integ_test.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/storage/storage_integ_test.go]: Can remove these types of imports. Should be handled by the sqorc.Open, or somewhere along those lines

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go]: Just `if err == merrors.ErrNotFound`

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go]: Do this block in a helper fn

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go]: Can we make a single batch call to GetStates?

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go]: Unsure: maybe it makes sense to just return `union (as a set) -> array -> sort` ?

Would be a more stable output, and this helper fn's impl would be simpler

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go]: fyi can do `map[string]struct[}{}` to take up less space, but it's perhaps less clear to understand

File: orc8r/cloud/configs/service_registry.yml
Comment (hcgatewood) [inline orc8r/cloud/configs/service_registry.yml:45]: First time we're incrementing an existing indexer's version 🙂

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Looks good. Couple small things then I'll be content

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (karthiksubraveti) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go:117]: minor nit: fix incorrect function name

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (karthiksubraveti) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go:165]: Why are we sorting them ?

File: orc8r/cloud/go/services/directoryd/storage/storage_blobstore.go
Comment (karthiksubraveti) [inline orc8r/cloud/go/services/directoryd/storage/storage_blobstore.go:151]: minor nit: Can you add some minor debug prints in case we want to debug any issue with greater verbosity?

**Review by karthiksubraveti — COMMENTED** [review]
Most changes LGTM. Will let Hunter approve it since he has already reviewed the changes. Wondering if there is a following API changes as well to return this mapping ?

File: orc8r/cloud/go/services/directoryd/client_api.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/client_api.go:163]: Nit for here and elsewhere: the point of the `Get*` methods is to handle if the receiver is nil. We should only be using this construction if it's actually expected that the receiver will be nil -- otherwise it's confusing / miscommunicates to the reader

File: orc8r/cloud/go/services/directoryd/client_api.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/client_api.go:161]: errors.Wrapf

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go:101]: if -> If

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go:138]: since -> Since

File: orc8r/cloud/go/services/directoryd/servicers/servicer.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/servicers/servicer.go:165]: Add comment for why sorting

Assuming it's for determinism, which is fine

File: orc8r/cloud/go/services/directoryd/types/types.go
Comment (hcgatewood) [inline orc8r/cloud/go/services/directoryd/types/types.go:88]: Bump on this `(string, bool)` comment

File: orc8r/protos/directoryd.proto
Comment (hcgatewood) [inline orc8r/protos/directoryd.proto:85]: `returns (MapHWIDToDirectoryRecordIDsResponse`, with empty response type

**Review by hcgatewood — APPROVED** [review]
Couple nits. Probably need to rebase before merging

**Review by hcgatewood — APPROVED** [review]
@mpgermano ready to merge?

**Comment by ekowtaylor** [comment]
Once this PR has been approved and passing all CI checks, use the ready2merge label to indicate the maintainers can merge your PR.

**Comment by hcgatewood** [comment]
@ekowtaylor just stumbled upon this PR. (a) Do we have any metrics or alerting set up to detect outdated PRs like this? And (b) do we still need this functionality or should I close the PR (it's out of date and would require a good bit of work to rectify with current master, along with the subsequent PR that Michael hadn't pushed yet)

**Comment by ekowtaylor** [comment]
@hcgatewood there is no tooling for this yet, but I usually use a month as a good rule of thumb but before closing it, I try to reach out to the author if they want to update it first.

**Comment by hcgatewood** [comment]
Closing due to inactivity

**Pull request closed without merging**
